File: java/src/main/java/dxjruby/Window.java
package dxjruby;

import static dxjruby.util.Utils.toInt;

import java.awt.AlphaComposite;
import java.awt.Color;
import java.awt.Composite;
import java.awt.Graphics2D;
import java.awt.RenderingHints;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.awt.geom.AffineTransform;
import java.awt.image.BufferedImage;
import java.util.function.Consumer;

import javax.swing.JFrame;

import dxjruby.DrawQueue.Command;
import dxjruby.graphic.Drawer;

public class Window {

    private static int width = 640;
    private static int height = 480;
    private static Color bgcolor = new Color(0, 0, 0);
    private static MainPanel mainPanel;
    private static final JFrame frame;

    static {
        frame = new JFrame();
    }

    public static void startGui() {
        frame.setLocationByPlatform(true);

        frame.setResizable(false);
        setCloseOperation(frame);

        mainPanel = new MainPanel(
                getWidth(),
                getHeight(),
                getBgcolor()
                );

        frame.add(mainPanel);
        frame.pack();

        frame.setVisible(true);
    }

    private static void setCloseOperation(final JFrame frame) {
        frame.addWindowListener(new WindowAdapter() {
            public void windowClosing(final WindowEvent e) {
                System.exit(0);
            }
        });
    }

    public static void updateInputState() {
        final long tNow = System.nanoTime();
        Input.updateMouseState(tNow);
        Input.updateKeyState(tNow);
    }

    public static void requestPaint() {
        mainPanel.requestPaint();
    }

    public static String getCaption() {
        return frame.getTitle();
    }

    public static void setCaption(final String caption) {
        frame.setTitle(caption);
    }

    // --------------------------------

    public static void drawImage(
            final double x, final double y,
            final Image image,
            final int z
            ) {
        addToDrawQueue(
                z,
                g2 -> {
                    final BufferedImage img = image.getAwtImage();
                    g2.drawImage(img, toInt(x), toInt(y), null);
                });
    }

    public static void drawEx(
            final double x, final double y,
            final Image image,
            final int z,
            final double cx, final double cy,
            final double scalex, final double scaley,
            final double angle,
            final int alpha
            ) {
        addToDrawQueue(
                z,
                g2 -> {
                    final AffineTransform atReset = g2.getTransform();
                    final Composite compReset = g2.getComposite();

                    final AffineTransform atTranslateRev =
                            AffineTransform.getTranslateInstance(
                                    (x + cx),
                                    (y + cy)
                            );

                    final AffineTransform atTranslate =
                            AffineTransform.getTranslateInstance(
                                    -(x + cx),
                                    -(y + cy)
                            );

                    final AffineTransform atRotate =
                            AffineTransform.getRotateInstance(
                                    Math.toRadians(angle)
                            );

                    final AffineTransform atScale =
                            AffineTransform.getScaleInstance(
                                    scalex, scaley
                            );

                    final AffineTransform at = new AffineTransform();
                    at.concatenate(atTranslateRev);
                    at.concatenate(atRotate);
                    at.concatenate(atScale);
                    at.concatenate(atTranslate);
                    g2.setTransform(at);

                    final float fAlpha = Integer.valueOf(alpha).floatValue() / 255.0F;
                    g2.setComposite(AlphaComposite.getInstance(AlphaComposite.SRC_OVER, fAlpha));

                    g2.drawImage(image.getAwtImage(), toInt(x), toInt(y), null);

                    g2.setTransform(atReset);
                    g2.setComposite(compReset);
                });
    }

    public static void drawFont(
            final double x, final double y, final String text,
            final dxjruby.Font font,
            // options
            final int z,
            final Color color
            ) {
        addToDrawQueue(
                z,
                g2 -> {
                    g2.setColor(color);

                    g2.setRenderingHint(
                            RenderingHints.KEY_TEXT_ANTIALIASING,
                            RenderingHints.VALUE_TEXT_ANTIALIAS_ON);

                    final java.awt.Font awtFont = font.getAwtFont();
                    int fontHeight = g2.getFontMetrics(awtFont).getHeight();
                    g2.setFont(awtFont);
                    g2.drawString(text, toInt(x), toInt(y) + fontHeight);
                });
    }

    public static void drawLine(
            final double x1, final double y1,
            final double x2, final double y2,
            final Color c, final int z
            ) {
        addToDrawQueue(
                z,
                g2 -> Drawer.line(g2, x1, y1, x2, y2, c)
                );
    }

    public static void drawBox(
            final double x1,
            final double y1,
            final double x2,
            final double y2,
            final Color color,
            final int z
            ) {
        addToDrawQueue(
                z,
                g2 -> Drawer.box(g2, x1, y1, x2, y2, color)
                );
    }

    public static void drawBoxFill(
            final double x1,
            final double y1,
            final double x2,
            final double y2,
            final Color color,
            final int z
            ) {
        addToDrawQueue(
                z,
                g2 -> Drawer.boxFill(g2, x1, y1, x2, y2, color)
                );
    }

    public static void drawCircle(
            final double x, final double y, final double r,
            final Color c, final int z
            ) {
        addToDrawQueue(
                z,
                g2 -> Drawer.circle(g2, x, y, r, c)
                );
    }

    public static void drawCircleFill(
            final double x, final double y, final double r,
            final Color c, final int z
            ) {
        addToDrawQueue(
                z,
                g2 -> Drawer.circleFill(g2, x, y, r, c)
                );
    }

    private static void addToDrawQueue(final int z, final Consumer<Graphics2D> func) {
        DrawQueue.add(z, new Command(func));
    }

    // --------------------------------

    public static int getWidth() {
        return width;
    }

    public static void setWidth(final int width) {
        Window.width = width;
    }

    public static int getHeight() {
        return height;
    }

    public static void setHeight(final int height) {
        Window.height = height;
    }

    public static Color getBgcolor() {
        return bgcolor;
    }

    public static void setBgcolor(final Color color) {
        Window.bgcolor = color;
    }

}

File: lib/dxjruby/window.rb
require_relative "remote_resource"

module DXJRuby
  module Window
    # Load resources specified with Image.register or Sound.register
    # Call block when loaded
    def self.load_resources(&block)
      RemoteResource._load_resources
      block.call
    end

    class FpsManager
      attr_accessor :count
      attr_accessor :delta
      attr_accessor :real_fps
      attr_accessor :span_frame_acc
      attr_reader :span_per_frame
      attr_reader :fps

      def initialize
        @count = 0
        @delta = 0.0 # sec
        @real_fps = nil
        @span_frame_acc = 0.0 # sec accumurated
        @fps = 60
        change_fps(@fps)
      end

      def change_fps(fps)
        @fps = fps
        @real_fps = @fps
        @span_per_frame = 1.0 / @fps
      end
    end

    @@fpsm = FpsManager.new

    # Start main loop
    #
    # TODO When called twice, previous loop is stopped (this is useful
    #      when implementing interactive game editor, etc.)
    def self.loop(&block)
      if j_Window.get_caption() == ""
        j_Window.set_caption("DXJRubyApplication")
      end
      j_Window.start_gui()

      t_base_loop = Time.now
      t_base_frame = Time.now

      Kernel.loop do
        t_now = Time.now
        @@fpsm.delta += (t_now - t_base_loop) / @@fpsm.span_per_frame
        t_base_loop = t_now

        if @@fpsm.delta >= 1.0
          while @@fpsm.delta >= 1.0
            @@fpsm.delta -= 1.0
          end

          @@fpsm.count += 1

          j_Window.update_input_state()
          block.call

          j_Window.request_paint()

          t_now = Time.now
          span_delta = t_now - t_base_frame
          t_base_frame = t_now
          @@fpsm.span_frame_acc += span_delta

          if @@fpsm.span_frame_acc >= 1.0
            @@fpsm.span_frame_acc = 0.0
            @@fpsm.real_fps = @@fpsm.count
            @@fpsm.count = 0
          end
        end
      end
    end

    ## # (DXOpal original) Pause & resume
    ## def self.pause
    ## def self.paused?
    ## def self.resume
    ## def self.draw_pause_screen

    ## # (internal) call @@block periodically
    ## def self._loop(timestamp)

    ## def self._init

    ## # Return internal DXOpal::Image object (for experimental/hacking use)
    ## def self._img

    def self.fps
      @@fpsm.fps
    end

    def self.fps=(fps)
      @@fpsm.change_fps(fps)
    end

    def self.real_fps
      @@fpsm.real_fps
    end

    def self.width
      j_Window.get_width()
    end

    # Set window width and resize the canvas
    # Set `nil` to maximize canvas
    # TODO resize, maximize
    def self.width=(w)
      j_Window.set_width(w)
    end

    def self.height
      j_Window.get_height()
    end

    # Set window height and resize the canvas
    # Set `nil` to maximize canvas
    # TODO resize, maximize
    def self.height=(h)
      j_Window.set_height(h)
    end

    ## @@bgcolor = Constants::Colors::C_BLACK
    ## def self.bgcolor
    def self.bgcolor=(col)
      j_Window.set_bgcolor(j_color(col))
    end

    def self.draw(x, y, image, z=0)
      j_Window.draw_image(x, y, image.to_j, z)
    end

    ## def self.draw_scale(x, y, image, scale_x, scale_y, center_x=nil, center_y=nil, z=0)

    ## def self.draw_rot(x, y, image, angle, center_x=nil, center_y=nil, z=0)

    # blend is not supported
    def self.draw_ex(x, y, image, options={})
      z = options[:z] || 0
      cx = options[:center_x] || (image.width.to_f / 2)
      cy = options[:center_y] || (image.height.to_f / 2)
      scalex = options[:scale_x] || 1.0
      scaley = options[:scale_y] || 1.0
      angle = options[:angle] || 0.0
      alpha = options[:alpha] || 255

      j_Window.draw_ex(
        x, y, image.to_j,
        # options
        z,
        cx.to_f, cy.to_f,
        scalex.to_f, scaley.to_f,
        angle.to_f,
        alpha
      )
    end

    def self.draw_font(x, y, string, font, option={})
      unless string.is_a?(String)
        raise TypeError.new("wrong argument type #{string.class} (expected String)")
      end

      z = option[:z] || 0
      color = option[:color] || [255, 255, 255]

      j_Window.draw_font(
        x, y, string, font.j_font,
        # options
        z, j_color(color)
      )
    end

    ## def self.draw_pixel(x, y, color, z=0)

    def self.draw_line(x1, y1, x2, y2, color, z=0)
      j_Window.draw_line(x1, y1, x2, y2, j_color(color), z)
    end

    def self.draw_box(x1, y1, x2, y2, color, z=0)
      j_Window.draw_box(x1, y1, x2, y2, j_color(color), z)
    end

    def self.draw_box_fill(x1, y1, x2, y2, color, z=0)
      j_Window.draw_box_fill(x1, y1, x2, y2, j_color(color), z)
    end

    def self.draw_circle(x, y, r, color, z=0)
      j_Window.draw_circle(x, y, r, j_color(color), z)
    end

    def self.draw_circle_fill(x, y, r, color, z=0)
      j_Window.draw_circle_fill(x, y, r, j_color(color), z)
    end

    ## def self.draw_triangle(x1, y1, x2, y2, x3, y3, color, z=0)

    ## def self.draw_triangle_fill(x1, y1, x2, y2, x3, y3, color, z=0)

    ## # (internal)
    ## def self.enqueue_draw(z, *args)

    #
    # DXRuby
    #

    def self.caption
      j_Window.get_caption()
    end

    def self.caption=(title)
      unless title.is_a?(String)
        raise TypeError.new("wrong argument type #{title.class} (expected String)")
      end

      j_Window.set_caption(title)
    end

    private

    def self.j_color(color_array)
      Java::dxjruby.DXJRuby.to_awt_color(*color_array)
    end

    def self.j_Window
      Java::dxjruby.Window
    end
  end
end
